Expression function implementation for aggregates

Nathan-Fenner opened the pull request:
This PR implements calls to the 4 implemented aggregation functions.

It also performs ancillary cleanup for arithmetic calls.

Nathan-Fenner commented on query/expression.go:
```diff
@@ -144,19 +144,66 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
-			func(left, right float64) float64 { return left * right })
+		fallthrough
+	case "/":
+		arguments := make([]value, len(expr.arguments))
```
Although it's possible to move this definition up to the top of the structure (especially since it's repeated for the aggregate functions), in order to change the `Timerange` to accommodate time shifting, arguments can't be evaluated until the function name is known (and then they may need to be evaluated in a special way).

achow commented on query/expression.go:
```diff
@@ -144,19 +144,66 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
+		if err != nil {
+			return nil, err
+		}
+		return seriesListValue(aggregateBy(list, aggregateMap[funMap[name]], expr.groupBy)), nil
```
`aggregateMap` here is really confusing since it appears to come out of nowhere. The `*Aggregate` enums also do the same... Maybe we should move these function implementations to their own package or something? How things are set up makes some of these references a little too magic for my liking, though maybe that's just how go works.

@jeeyoungk

Nathan-Fenner replied:
one solution might be to replace the `aggregateMap` in `aggregate.go` (which is of type `map[aggregate]func([]float64) float64`) into the map defined here directly, giving it the new type `map[string]func([]float64) float64` and eliminating the `Aggregate` enum...

There needs to be some mapping from `String` to `func([]float64) float64` somewhere, though it wouldn't need to necessarily be a map (it could also be a list of structs or something similar, each a name-function pair).

achow replied:
Imo, callers should only be concerned with the enum, and the callees should do the mappings from enums to implementation. I wonder if there's a better way in general to do dispatch though.

Nathan-Fenner replied:
There was some previous discussion: https://github.com/square/metrics/pull/35#discussion_r31552407

What should be done here? Currently there is a "double" lookup, converting the `string` to the `Aggregate` through the local-variable `funMap`, and then again through `aggregateMap` (defined in `aggregate.go`).

The lookup from `enum => aggregate` could be done inside the `aggregateBy` function instead.

achow replied:
Yes, have the enum => aggregate mapping occur in `aggregateBy`.

I still think we should export the enums and `aggregateBy`, not for package-external visibility, but for package-internal identification of things that are meant to be used. Use of unexported things should be limited to the files in which they are defined. Though, if we start doing this it _may_ be useful to start splitting these into packages.

Nathan-Fenner replied:
How should we convert the function name (string) into the associated enum?

Should aggregate just define a `map[string]aggregate` and callers look up their given function name before calling `aggregateBy`?

Nathan-Fenner replied:
I exported `AggregateBy` and also `AggregateMap` but I'm not sure about that one, and also made `AggregateBy` do the lookup (`aggregate => func`) itself.

Nathan-Fenner commented:
I moved `aggregate.go` and `aggregate_test.go` into a new package `query/aggregate`.

In addition, I changed the interface of `AggregateBy` from:

`func AggregateBy(list api.SeriesList, aggregator aggregate, tags []string) api.SeriesList`

to:

`func AggregateBy(list api.SeriesList, name string, tags []string) (api.SeriesList, error)`

The principle change is that an enum is no longer used to decide which aggregator is run- instead, a string is used.

The reason for this change is that the use of an enum provides little safety over pure string keys, the reason for this being that AggregateBy is *always* called only using user-supplied data (the function name) and never by the code itself, where verifying that the enum actually exists is possible.

Since users always supplied the data in the form of a string, with enums, it was necessary to perform two transformations: (`string => enum type`) and then (`enum type => function`). However, since the intermediate result is never used, except to be immediately transformed into the function by looking it up in the map, the intermediate result is not particularly useful.

`AggregateBy` now instead performs the lookup itself using the string, which means only one table of `name` / `func` needs to be defined (it's currently the unexported name `aggregateMap`). If the given name is not present in the table, then `AggregateBy` returns an error.

jeeyoungk commented on query/expression.go:
```diff
@@ -144,19 +145,52 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
+
+	name := expr.functionName
+	switch name {
+	case "+", "-", "*", "/":
```
we should pull out `operatorMap` out, and have if statement on whether the value belongs on the `operatorMap` - this ensure that we don't have to manually keep `case` and `operatorMap` in sync.

Nathan-Fenner replied:
The `switch`/`case` has been replaced by `value, ok :=` lookups in hashmaps (through a function `GetAggregate` for aggregation functions)

jeeyoungk commented on query/expression.go:
```diff
@@ -144,19 +145,52 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
+			"/": func(x, y float64) float64 { return x / y },
+		}
+		return evaluateBinaryOperation(context, name, left, right, operatorMap[name])
+	case "aggregate.sum", "aggregate.mean", "aggregate.min", "aggregate.max":
```
same as here - i think the `case` statement should use values from some map.

jeeyoungk commented on query/aggregate/aggregate.go:
```diff
@@ -72,17 +74,8 @@ func groupBy(list api.SeriesList, tags []string) []group {
-
-var aggregateMap = map[aggregate]func([]float64) float64{
-	sumAggregate:
+var aggregateMap = map[string]func([]float64) float64{
```
we can make this code more "modular" by instead of hard-coding `map[name]function`, expose a function called `NewAggregate(name string, func function)` which can be used to add functions. In that case, you can define these aggregate functions in different party, and you can compose them that way.

Nathan-Fenner replied:
`aggregateMap` is now accessed through `GetAggregate` and `NewAggregate` functions; `GetAggregate` returns `func([]float64) float64, bool` for `ok` checks, and `NewAggregate` takes a name and a function.

jeeyoungk commented on query/aggregate/aggregate.go:
```diff
@@ -133,6 +123,17 @@ var aggregateMap = map[aggregate]func([]float64) float64{
+
+// AddAggregate adds an aggregate of a given name to the aggregate map.
+func NewAggregate(name string, aggregate func([]float64) float64) {
+	aggregateMap[name] = aggregate
```
let's `panic` on a duplicate registration.

jeeyoungk commented:
LGTM with nit
